File: leetcode/881/881.py
# Given a string s, return the first non-repeating character in it and return its index. If it does not exist, return -1.

class Solution:
    def firstUniqChar(self, s: str) -> int:
        counts = {}
        
        # fill in dict with count of chars
        for c in s:
            if c in counts:
                counts[c] = counts[c] + 1
            else:
                counts[c] = 1
                
        # find first char with count of 1
        for i in range(len(s)):
            if counts[s[i]] == 1:
                return i
            
        return -1

File: leetcode/881/881.go
// Given a string s, find the first non-repeating character in it and return its index. If it does not exist, return -1.

func firstUniqChar(s string) int {
    m := buildMap(s)
    return findFirstUniq(s, m)
}

func buildMap(s string) map[byte]int {
    chars := make(map[byte]int)
    for i:= 0; i < len(s); i++ {
        count := chars[s[i]]
        count++
        chars[s[i]] = count
    }
    return chars
}

func findFirstUniq(s string, m map[byte]int) int{
    for i := 0; i < len(s); i++ {
        if m[s[i]] == 1 {
            return i
        }
    }
    return -1
}